timesync/peersync/sync.go

```diff
@@ -0,0 +1,330 @@
+package peersync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/p2pcrypto"
+	"github.com/spacemeshos/go-spacemesh/p2p/server"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	protocolName = "/peersync/1.0/"
+)
+
+var (
+	// ErrPeersNotSynced returned if system clock is out of sync with peers clock for configured period of time.
+	ErrPeersNotSynced = errors.New("timesync: peers are not time synced")
+	// ErrTimesyncFailed returned if we weren't able to collect enough clock samples from peers.
+	ErrTimesyncFailed = errors.New("timesync: failed request")
+)
+
+//go:generate mockgen -package=mocks -destination=./mocks/mocks.go -source=./sync.go Time,Network
+
+// Time ...
+type Time interface {
+	Now() time.Time
+}
+
+// Network ...
+type Network interface {
+	RegisterDirectProtocolWithChannel(protocol string, ingressChannel chan service.DirectMessage) chan service.DirectMessage
+	SendWrappedMessage(ctx context.Context, nodeID p2pcrypto.PublicKey, protocol string, payload *service.DataMsgWrapper) error
+	SubscribePeerEvents() (added, expired chan p2pcrypto.PublicKey)
+}
+
+type systemTime struct{}
+
+func (s systemTime) Now() time.Time {
+	return time.Now()
+}
+
+// Request for time from a peer.
+type Request struct {
+	ID uint64
+}
+
+// Response for time from a peer.
+type Response struct {
+	ID        uint64
+	Timestamp int64
+}
+
+// DefaultConfig for Sync.
+func DefaultConfig() Config {
+	return Config{
+		RoundRetryInterval:  5 * time.Second,
+		RoundInterval:       30 * time.Minute,
+		RoundTimeout:        5 * time.Second,
+		MaxClockOffset:      10 * time.Second,
+		MaxOffsetErrors:     10,
+		RequiredResponses:   3,
+		ResponsesBufferSize: 20,
+	}
+}
+
+// Config for Sync.
+type Config struct {
+	Disable            bool          `mapstructure:"disable"`
+	RoundRetryInterval time.Duration `mapstructure:"round-retry-interval"`
```

Comment:
add comments about time units

Reply:
it allows to specify any time.Duration e.g 5s/10m/6h